astropy/stats/sigma_clipping.py

```diff
@@ -292,45 +293,44 @@ def _sigmaclip_fast(self, data, axis=None,
         Fast C implementation for simple use cases
         """
 
-        # The Cython implementation takes 2-d arrays and assumes axis=0, so we need
-        # to normalize the input array so that we can treat it in this way. The Cython
-        # routine produces a mask with the same 2-d shape, so we pre-allocate the
-        # mask here and return a 2-d view.
+        # The gufunc implementation assumes axis=-1, so we need to normalize
+        # the input array so that we can treat it in this way.
+        # TODO: for single axis, can just pass that on to gufunc!
 
-        reshape_1d = axis is None
+        if axis is None:
+            axis = -1 if data.ndim == 1 else tuple(range(data.ndim))
 
-        if reshape_1d:
-            axis = tuple(range(data.ndim))
-            data = data.reshape(data.shape + (1,))
+        if not isiterable(axis):
+            axis = normalize_axis_index(axis, data.ndim)
+            data_reshaped = data.astype(float, copy=False)
         else:
-            if not isiterable(axis):
-                axis = (axis,)
-            axis = tuple(data.ndim + n if n < 0 else n for n in axis)
-
-        data_reordered = _move_tuple_axes_first(data, axis)
-        data_2d = data_reordered.reshape((data_reordered.shape[0], -1))
+            axis = tuple(normalize_axis_index(ax, data.ndim) for ax in axis)
+            transposed_axes = tuple(ax for ax in range(data.ndim)
+                                    if ax not in axis) + axis
+            data_transposed = data.transpose(transposed_axes)
+            transposed_shape = data_transposed.shape
+            data_reshaped = data_transposed.astype(float, copy=False).reshape(
+                transposed_shape[:data.ndim-len(axis)]+(-1,))
+            axis = -1
 
-        mask = ~np.isfinite(data)
-        mask_reordered = _move_tuple_axes_first(mask, axis)
-        mask_2d = mask_reordered.reshape(data_2d.shape)
+        mask = ~np.isfinite(data_reshaped)
 
         if np.any(mask):
             warnings.warn('Input data contains invalid values (NaNs or '
                           'infs), which were automatically clipped.',
                           AstropyUserWarning)
-
-        bounds = _sigma_clip_fast(data_2d, mask_2d, 0 if self.cenfunc == 'mean' else 1,
+        bounds = _sigma_clip_fast(data_reshaped, mask, self.cenfunc != 'mean',
                                   -1 if np.isinf(self.maxiters) else self.maxiters,
-                                  self.sigma_lower, self.sigma_upper)
-
-        mask_2d |= data_2d < bounds[0]
-        mask_2d |= data_2d > bounds[1]
+                                  self.sigma_lower, self.sigma_upper, axis=axis)
 
-        # TODO: reshape bounds for 3d+ cases
+        mask |= data_reshaped < np.expand_dims(bounds[0], axis)
+        mask |= data_reshaped > np.expand_dims(bounds[1], axis)
```

Comment:
Is this easy to do in the gufunc itself? Ideally I wanted to do it there to avoid any additional memory allocation but it was difficult with my previous approach.

Reply:
Yes, one could add a mask as output and one could do it inplace by giving the same mask as output in the gufunc call.